### PR TITLE
ESP32: Address -Wundef build errors and enable it for platform

### LINKED ISSUE #29586: [ESP32] Various -Wundef compile issues in src/platform/ESP32/CHIPPlatformConfig.h
### Reproduction steps

https://github.com/project-chip/connectedhomeip/actions/runs/6419947072/job/17430925276?pr=29582 (from turning on `-Wundef` in our global compile options).

It looks like `src/platform/ESP32/CHIPPlatformConfig.h` tests various macros without including the things that would define them.  If that is expected, it should be testing `defined()` or `#ifdef`:

* CONFIG_ENABLE_ICD_SERVER
* CONFIG_SECURITY_TEST_MODE

```
INFO    ../../../../config/esp32/third_party/connectedhomeip/src/platform/ESP32/CHIPPlatformConfig.h:79:5: error: "CONFIG_ENABLE_ICD_SERVER" is not defined, evaluates to 0 [-Werror=undef]
INFO       79 | #if CONFIG_ENABLE_ICD_SERVER
INFO          |     ^~~~~~~~~~~~~~~~~~~~~~~~
INFO    ../../../../config/esp32/third_party/connectedhomeip/src/platform/ESP32/CHIPPlatformConfig.h:61:40: error: "CONFIG_SECURITY_TEST_MODE" is not defined, evaluates to 0 [-Werror=undef]
INFO       61 | #define CHIP_CONFIG_SECURITY_TEST_MODE CONFIG_SECURITY_TEST_MODE
INFO          |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~
```

@dhrishi @shubhamdp 

### Platform

esp32

### Platform Version(s)

_No response_

### Type

CI tested

### (Optional) If manually tested please explain why this is only manually tested

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shubhamdp**
Fixes https://github.com/project-chip/connectedhomeip/issues/29586

- Fixed the build errors reported after enabling `-Wundef` and enabled it in ESP32 paltform.

Defining CHIP_SYSTEM_CONFIG_LWIP_PBUF_FROM_CUSTOM_POOL

Some platforms do not use the lwip from third_party/lwip and may not have LWIP_PBUF_FROM_CUSTOM_POOLS defined in lwip/opt.h. When building apps with -Wundef cflag enabled, it may fail to find LWIP_PBUF_FROM_CUSTOM_POOLS hence this option. This defaults to LWIP_PBUF_FROM_CUSTOM_POOLS if defined, otherwise 0.

### CODE REVIEW

File: src/messaging/ReliableMessageProtocolConfig.h [hunk 1]
```diff
@@ -129,7 +129,7 @@ namespace chip {
 #ifndef CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE
 #if CHIP_SYSTEM_CONFIG_USE_LWIP
 
-#if !LWIP_PBUF_FROM_CUSTOM_POOLS && PBUF_POOL_SIZE != 0
+#if !(defined(LWIP_PBUF_FROM_CUSTOM_POOLS) && LWIP_PBUF_FROM_CUSTOM_POOLS) && PBUF_POOL_SIZE != 0
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageProtocolConfig.h:132]: This is not right.  `#include <lwip/opt.h>` above pulls in a definition of LWIP_PBUF_FROM_CUSTOM_POOLS, no?

File: src/platform/ESP32/CHIPDevicePlatformConfig.h [hunk 2]
```diff
@@ -48,39 +48,53 @@
...
+#endif // CONFIG_ENABLE_ETHERNET_TELEMETRY
 
-#if CONFIG_ENABLE_ICD_SERVER
+#if defined(CONFIG_ENABLE_ICD_SERVER) && CONFIG_ENABLE_ICD_SERVER
 #define CHIP_DEVICE_CONFIG_ICD_SLOW_POLL_INTERVAL chip::System::Clock::Milliseconds32(CONFIG_ICD_SLOW_POLL_INTERVAL_MS)
 #define CHIP_DEVICE_CONFIG_ICD_FAST_POLL_INTERVAL chip::System::Clock::Milliseconds32(CONFIG_ICD_FAST_POLL_INTERVAL_MS)
-#endif // CHIP_CONFIG_ENABLE_ICD_SERVER
+#endif
 
+#if defined(CONFIG_ENABLE_CHIPOBLE) && CONFIG_ENABLE_CHIPOBLE
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/CHIPDevicePlatformConfig.h:86]: If CONFIG_ENABLE_CHIPOBLE is defined and false, wouldn't you still want to set CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE false here?  This part does not make sense to me.  Similar for a bunch of the checks below....

In general, if these CONFIG_ values are just "on or off" and "on" always means "defined" and "off" means "not defined", we should be checking ifdef only, not the value, I would think.

File: src/platform/ESP32/ConfigurationManagerImpl.cpp [hunk 3]
```diff
@@ -31,9 +31,9 @@
 #include <platform/ESP32/ESP32Config.h>
 #include <platform/internal/GenericConfigurationManagerImpl.ipp>
 
-#if CHIP_DEVICE_CONFIG_ENABLE_ETHERNET
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_ETHERNET) && CHIP_DEVICE_CONFIG_ENABLE_ETHERNET
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/ConfigurationManagerImpl.cpp:34]: This is wrong.  CHIP_DEVICE_CONFIG_ENABLE_ETHERNET should always be defined in the device config.  If that's not happening, either someone messed up, or (worse) this file is not including the right config values.

I see that src/platform/ESP32/CHIPDevicePlatformConfig.h defines CHIP_DEVICE_CONFIG_ENABLE_ETHERNET only sometimes; should it be defining it to 0 in other cases, if this is an esp32-specific define?  If this is a general define, cross-platform, then CHIPDeviceConfig.h should be defaulting it, I assume.

File: src/system/SystemPacketBufferInternal.h [hunk 4]
```diff
@@ -82,7 +82,7 @@
  *
  * True if packet buffers are allocated from an LwIP custom pool.
  */
-#if CHIP_SYSTEM_PACKETBUFFER_FROM_LWIP_POOL && !LWIP_PBUF_FROM_CUSTOM_POOLS
+#if CHIP_SYSTEM_PACKETBUFFER_FROM_LWIP_POOL && !(defined(LWIP_PBUF_FROM_CUSTOM_POOLS) && LWIP_PBUF_FROM_CUSTOM_POOLS)
```
Comment (bzbarsky-apple) [inline src/system/SystemPacketBufferInternal.h:85]: Again, this is wrong.  The `#include <lwip/opt.h>` above should be pulling in those values.  This is breaking detection of cases when configs are not correctly set up.

File: src/system/SystemStats.cpp [hunk 5]
```diff
@@ -38,7 +38,7 @@ namespace System {
 namespace Stats {
 
 static const Label sStatsStrings[chip::System::Stats::kNumEntries] = {
-#if CHIP_SYSTEM_CONFIG_USE_LWIP && LWIP_PBUF_FROM_CUSTOM_POOLS
+#if CHIP_SYSTEM_CONFIG_USE_LWIP && defined(LWIP_PBUF_FROM_CUSTOM_POOLS) && LWIP_PBUF_FROM_CUSTOM_POOLS
```
Comment (bzbarsky-apple) [inline src/system/SystemStats.cpp:41]: As above: this is wrong, though this file does not `#include <lwip/opt.h>` and should.

File: src/system/SystemStats.h [hunk 6]
```diff
@@ -52,7 +52,7 @@ namespace Stats {
 
 enum
 {
-#if CHIP_SYSTEM_CONFIG_USE_LWIP && LWIP_PBUF_FROM_CUSTOM_POOLS
+#if CHIP_SYSTEM_CONFIG_USE_LWIP && defined(LWIP_PBUF_FROM_CUSTOM_POOLS) && LWIP_PBUF_FROM_CUSTOM_POOLS
```
Comment (bzbarsky-apple) [inline src/system/SystemStats.h:55]: As above.  And this file _does_ `#include <lwip/opt.h>`.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/messaging/ReliableMessageProtocolConfig.h (hunk 1 above)
Comment (shubhamdp) [reply src/messaging/ReliableMessageProtocolConfig.h:132]: `#include <lwip/opt.h>` includes `lwipopts.h` and ESP32 uses the lwip from the esp-idf which does not contains the config `LWIP_PBUF_FROM_CUSTOM_POOLS`.

`third_party/lwip/repo/lwip` was added two years back and the one in esp-idf is pretty new so, this option could have been removed in newer lwip.

File: src/platform/ESP32/CHIPDevicePlatformConfig.h (hunk 2 above)
Comment (shubhamdp) [reply src/platform/ESP32/CHIPDevicePlatformConfig.h]: Fixed setting of `CHIP_*` configs and always setting them even if the configs from the ESP32 is undefined.

> In general, if these CONFIG_ values are just "on or off" and "on" always means "defined" and "off" means "not defined", we should be checking ifdef only, not the value, I would think.

You are right here, just performing an `#ifdef` check.

File: src/platform/ESP32/ConfigurationManagerImpl.cpp (hunk 3 above)
Comment (shubhamdp) [reply src/platform/ESP32/ConfigurationManagerImpl.cpp]: Fixed this

File: src/messaging/ReliableMessageProtocolConfig.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageProtocolConfig.h:132]: Ok, well, we need to figure out how to handle this case so we detect failure to include lwip/opt.h...

Is there a new thing that replaced LWIP_PBUF_FROM_CUSTOM_POOLS?  Is there something smart we can do about checking for another define that would always come from there assuming that it's included?

Comment (shubhamdp) [reply src/messaging/ReliableMessageProtocolConfig.h]: I have added a new config option `CHIP_SYSTEM_CONFIG_LWIP_PBUF_FROM_CUSTOM_POOL` which checks if `LWIP_PBUF_FROM_CUSTOM_POOLS` is present and if not, it defines it with value `0`.

File: src/system/SystemConfig.h [hunk 7]
```diff
@@ -771,3 +771,20 @@ struct LwIPEvent;
...
+ * @def CHIP_SYSTEM_CONFIG_LWIP_PBUF_FROM_CUSTOM_POOL
+ *
+ * @brief
+ *     Enable the use of lwIP pbufs from a custom pool
+ *
+ * This config option exist because not all platforms defines LWIP_PBUF_FROM_CUSTOM_POOLS in lwip/opt.h
+ * Defaults to LWIP_PBUF_FROM_CUSTOM_POOLS if defined, otherwise 0
+ */
+#ifndef CHIP_SYSTEM_CONFIG_LWIP_PBUF_FROM_CUSTOM_POOL
+#if defined(LWIP_PBUF_FROM_CUSTOM_POOLS)
```
Comment (bzbarsky-apple) [inline src/system/SystemConfig.h:785]: For this to work, doesn't this file need to include `<lwip/opt.h>` as needed?  This seems to be depending on include order, which we do not want.

Also, if we do this, then the lwip/opt.h includes that were added for uses of LWIP_PBUF_FROM_CUSTOM_POOLS should be removed in the places that are including this header and using CHIP_SYSTEM_CONFIG_LWIP_PBUF_FROM_CUSTOM_POOL instead.

Comment (shubhamdp) [reply src/system/SystemConfig.h]: This file already includes the `<lwip/opt.h>`, I moved newly added code just below that for easier reading. Also, addressed the other suggestions.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by wqx6 — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Pull request merged**